File: go/anyonce/engine.go
package anyonce

import (
	"context"
	"fmt"
	"sync/atomic"
	"time"
)

// StoreErrorMode is D13: fail-closed (default) or fail-open.
type StoreErrorMode string

const (
	FailClosed StoreErrorMode = "fail-closed"
	FailOpen   StoreErrorMode = "fail-open"
)

// Hooks are observability callbacks. A panicking hook is recovered and counted; it never reaches the engine.
type Hooks struct {
	OnAcquired   func(op Operation)
	OnReplayed   func(op Operation, record *Record)
	OnConflict   func(op Operation, leaseUntil time.Time)
	OnMismatch   func(op Operation, record *Record)
	OnStoreError func(op Operation, err error)
}

// Policy mirrors the TypeScript ExecutePolicy (requirements 3.3). Zero values take the defaults: a Lease, TTL or
// MaxResultBytes of zero or less becomes the DefaultPolicy value, a nil StoreResult becomes DefaultStoreResult, and an
// empty OnStoreError becomes FailClosed (D13).
// Execute normalizes its own copy, so the caller's Policy value is never modified.
type Policy struct {
	Lease          time.Duration
	TTL            time.Duration
	MaxResultBytes int
	StoreResult    func(StoredResult) bool
	OnStoreError   StoreErrorMode
	Clock          func() time.Time
	Hooks          Hooks
	HookErrors     *atomic.Int64
}

// ResultKind mirrors the TypeScript ExecuteResult union tag.
type ResultKind string

const (
	ResultExecuted   ResultKind = "executed"
	ResultReplayed   ResultKind = "replayed"
	ResultConflict   ResultKind = "conflict"
	ResultMismatch   ResultKind = "mismatch"
	ResultStoreError ResultKind = "store_error"
)

// Result is what Execute returns. Kind says which fields are meaningful: Result and Stored for executed, Record for
// replayed and mismatch, LeaseUntil for conflict, Err for store_error.
type Result struct {
	Kind       ResultKind
	Result     StoredResult
	Stored     bool
	Record     *Record
	LeaseUntil time.Time
	Err        error
}

// DefaultStoreResult is D6: store message outcomes and HTTP results below 500. A zero Status counts as 500.
func DefaultStoreResult(r StoredResult) bool {
	if r.Kind == KindMessage {
		return true
	}
	status := r.Status
	if status == 0 {
		status = 500
	}
	return status < 500
}

// DefaultPolicy returns the 3.3 defaults: 30 s lease, 24 h TTL, 1 MiB cap, fail-closed.
func DefaultPolicy() Policy {
	return Policy{Lease: 30 * time.Second, TTL: 24 * time.Hour, MaxResultBytes: 1 << 20, StoreResult: DefaultStoreResult, OnStoreError: FailClosed}
}

// ResultSize is the body length (Q17).
func ResultSize(r StoredResult) int { return len(r.Body) }

// OmitBody returns the D12 omitted form: status and headers kept, body dropped.
func OmitBody(r StoredResult) StoredResult {
	return StoredResult{Kind: r.Kind, Status: r.Status, Headers: r.Headers, Omitted: true}
}

func (p Policy) now() time.Time {
	if p.Clock != nil {
		return p.Clock()
	}
	return time.Now()
}

func (p Policy) safely(fn func()) {
	defer func() {
		if recovered := recover(); recovered != nil && p.HookErrors != nil {
			p.HookErrors.Add(1)
		}
	}()
	fn()
}

func (p Policy) storeError(op Operation, err error) {
	if p.Hooks.OnStoreError != nil {
		p.safely(func() { p.Hooks.OnStoreError(op, err) })
	}
}

func abandonQuietly(ctx context.Context, store Store, op Operation, fence int64, policy Policy) {
	if _, err := store.Abandon(ctx, op, fence); err != nil {
		policy.storeError(op, err)
	}
}

// Execute is the one state machine (requirements 3.3). It runs the handler at most once per acquired claim,
// replays completed results, and reports conflicts and mismatches through Result.Kind. Error contract (Q16):
// (Result, nil) for executed, replayed, conflict and mismatch; (Result{Kind: ResultStoreError}, err wrapping
// ErrStoreUnavailable) for a fail-closed store failure at begin; (Result{}, err wrapping the handler's error)
// when run fails, after abandoning the claim.
func Execute(ctx context.Context, store Store, op Operation, run func(ctx context.Context) (StoredResult, error), policy Policy) (Result, error) {
	if err := ctx.Err(); err != nil {
		return Result{}, err
	}
	defaults := DefaultPolicy()
	if policy.Lease <= 0 {
		policy.Lease = defaults.Lease
	}
	if policy.TTL <= 0 {
		policy.TTL = defaults.TTL
	}
	if policy.MaxResultBytes <= 0 {
		policy.MaxResultBytes = defaults.MaxResultBytes
	}
	if policy.StoreResult == nil {
		policy.StoreResult = defaults.StoreResult
	}
	if policy.OnStoreError == "" {
		policy.OnStoreError = defaults.OnStoreError
	}
	// Bookkeeping must outlive the caller's context. A client disconnect is the canonical retry trigger, and a
	// record left in flight because Complete or Abandon was refused would block that retry until the lease expires.
	bookkeeping := context.WithoutCancel(ctx)
	outcome, err := store.Begin(ctx, op, BeginOptions{Lease: policy.Lease, TTL: policy.TTL, Now: policy.now()})
	if err != nil {
		policy.storeError(op, err)
		if policy.OnStoreError == FailClosed {
			wrapped := fmt.Errorf("%w: begin: %w", ErrStoreUnavailable, err)
			return Result{Kind: ResultStoreError, Err: wrapped}, wrapped
		}
		result, runErr := run(ctx)
		if runErr != nil {
			return Result{}, fmt.Errorf("anyonce: handler failed: %w", runErr)
		}
		return Result{Kind: ResultExecuted, Result: result, Stored: false}, nil
	}

	switch outcome.Kind {
	case BeginCompleted:
		if policy.Hooks.OnReplayed != nil {
			policy.safely(func() { policy.Hooks.OnReplayed(op, outcome.Record) })
		}
		return Result{Kind: ResultReplayed, Record: outcome.Record}, nil
	case BeginInFlight:
		if policy.Hooks.OnConflict != nil {
			policy.safely(func() { policy.Hooks.OnConflict(op, outcome.LeaseUntil) })
		}
		return Result{Kind: ResultConflict, LeaseUntil: outcome.LeaseUntil}, nil
	case BeginMismatch:
		if policy.Hooks.OnMismatch != nil {
			policy.safely(func() { policy.Hooks.OnMismatch(op, outcome.Record) })
		}
		return Result{Kind: ResultMismatch, Record: outcome.Record}, nil
	}

	if policy.Hooks.OnAcquired != nil {
		policy.safely(func() { policy.Hooks.OnAcquired(op) })
	}
	result, runErr := run(ctx)
	if runErr != nil {
		abandonQuietly(bookkeeping, store, op, outcome.Fence, policy)
		return Result{}, fmt.Errorf("anyonce: handler failed: %w", runErr)
	}
	if !policy.StoreResult(result) {
		abandonQuietly(bookkeeping, store, op, outcome.Fence, policy)
		return Result{Kind: ResultExecuted, Result: result, Stored: false}, nil
	}
	payload := result
	if ResultSize(result) > policy.MaxResultBytes {
		payload = OmitBody(result)
	}
	status, err := store.Complete(bookkeeping, op, outcome.Fence, payload, policy.now())
	if err != nil {
		policy.storeError(op, err)
		return Result{Kind: ResultExecuted, Result: result, Stored: false}, nil
	}
	return Result{Kind: ResultExecuted, Result: result, Stored: status == CompleteOK}, nil
}

File: packages/core/src/engine.ts
import type {
  BeginOutcome,
  CompleteStatus,
  IdempotencyRecord,
  OmittedResult,
  Operation,
  Store,
  StoredResult,
} from './types';

export interface ExecuteHooks {
  onAcquired?(op: Operation): void;
  onReplayed?(op: Operation, record: IdempotencyRecord): void;
  onConflict?(op: Operation, leaseUntil: number): void;
  onMismatch?(op: Operation, record: IdempotencyRecord): void;
  onStoreError?(op: Operation, error: unknown): void;
}

export interface ExecutePolicy {
  leaseMs: number;
  ttlMs: number;
  maxResultBytes: number;
  storeResult: (result: StoredResult) => boolean;
  onStoreError: 'fail-closed' | 'fail-open';
  clock?: () => number;
  hooks?: ExecuteHooks;
  /** Incremented every time a hook throws. Hooks never throw into the engine (3.3). */
  hookErrors?: { count: number };
}

export type ExecuteResult =
  | { kind: 'executed'; result: StoredResult; stored: boolean }
  | { kind: 'replayed'; record: IdempotencyRecord }
  | { kind: 'conflict'; leaseUntil: number }
  | { kind: 'mismatch'; record: IdempotencyRecord }
  | { kind: 'store_error'; error: unknown };

export const DEFAULT_LEASE_MS = 30_000;
export const DEFAULT_TTL_MS = 86_400_000;
export const DEFAULT_MAX_RESULT_BYTES = 1_048_576;

/** D6: store message outcomes and HTTP results below 500. */
export function defaultStoreResult(result: StoredResult): boolean {
  return result.kind === 'message' || (result.status ?? 500) < 500;
}

export function defaultPolicy(overrides: Partial<ExecutePolicy> = {}): ExecutePolicy {
  return {
    leaseMs: DEFAULT_LEASE_MS,
    ttlMs: DEFAULT_TTL_MS,
    maxResultBytes: DEFAULT_MAX_RESULT_BYTES,
    storeResult: defaultStoreResult,
    onStoreError: 'fail-closed',
    ...overrides,
  };
}

/** Q17: the cap bounds stored bodies; headers are allowlisted and small. */
export function resultSize(result: StoredResult): number {
  return result.body?.byteLength ?? 0;
}

/** D12 and Q7: the omitted form keeps status and headers so a replay can reproduce them. */
export function omitBody(result: StoredResult): OmittedResult {
  const out: OmittedResult = { omitted: true, kind: result.kind };
  if (result.status !== undefined) out.status = result.status;
  if (result.headers !== undefined) out.headers = result.headers;
  return out;
}

type Safely = (fn: () => void) => void;

async function abandonQuietly(
  store: Store,
  op: Operation,
  fence: number,
  hooks: ExecuteHooks,
  safely: Safely,
): Promise<void> {
  try {
    await store.abandon(op, fence);
  } catch (error) {
    safely(() => hooks.onStoreError?.(op, error));
  }
}

/**
 * The one state machine (requirements 3.3). Runs the handler at most once per acquired claim, replays completed
 * results, reports conflicts and mismatches, and never lets a hook or a store failure change which of those it did.
 */
export async function execute(
  store: Store,
  op: Operation,
  /** Receives the fence of the acquired claim; 0 when running without a claim under fail-open. */
  run: (fence: number) => Promise<StoredResult>,
  policy: ExecutePolicy,
): Promise<ExecuteResult> {
  const now = (): number => (policy.clock ?? Date.now)();
  const hooks: ExecuteHooks = policy.hooks ?? {};
  const safely: Safely = (fn) => {
    try {
      fn();
    } catch {
      if (policy.hookErrors !== undefined) policy.hookErrors.count += 1;
    }
  };

  let outcome: BeginOutcome;
  try {
    outcome = await store.begin(op, { leaseMs: policy.leaseMs, ttlMs: policy.ttlMs, now: now() });
  } catch (error) {
    safely(() => hooks.onStoreError?.(op, error));
    if (policy.onStoreError === 'fail-closed') return { kind: 'store_error', error };
    return { kind: 'executed', result: await run(0), stored: false };
  }

  if (outcome.outcome === 'completed') {
    const { record } = outcome;
    safely(() => hooks.onReplayed?.(op, record));
    return { kind: 'replayed', record };
  }
  if (outcome.outcome === 'in_flight') {
    const { leaseUntil } = outcome;
    safely(() => hooks.onConflict?.(op, leaseUntil));
    return { kind: 'conflict', leaseUntil };
  }
  if (outcome.outcome === 'mismatch') {
    const { record } = outcome;
    safely(() => hooks.onMismatch?.(op, record));
    return { kind: 'mismatch', record };
  }

  const { fence } = outcome;
  safely(() => hooks.onAcquired?.(op));

  let result: StoredResult;
  try {
    result = await run(fence);
  } catch (error) {
    await abandonQuietly(store, op, fence, hooks, safely);
    throw error;
  }

  if (!policy.storeResult(result)) {
    await abandonQuietly(store, op, fence, hooks, safely);
    return { kind: 'executed', result, stored: false };
  }

  const payload = resultSize(result) > policy.maxResultBytes ? omitBody(result) : result;
  let status: CompleteStatus;
  try {
    status = await store.complete(op, fence, payload, now());
  } catch (error) {
    safely(() => hooks.onStoreError?.(op, error));
    return { kind: 'executed', result, stored: false };
  }
  return { kind: 'executed', result, stored: status === 'ok' };
}
